Add CheckpointTimer to simplify time measurement in loops

luator opened the pull request:
# Description
The `CheckpointTimer` class is intended for time measurement of loops
where the time spent in separate sections of the code is measured, as
well as the total duration of the loop.  See doxygen of the class for
more details.

It is basically a wrapper around a list of `Timer` instances with the
purpose to not clutter the measured code too much.  Also it can easily
be disabled via a template argument which should result in the compiler
optimizing away the function calls completely (i.e. no need to have
commented lines in the code).

# How I Tested
By using it in robot_interfaces::RobotBackend (see https://github.com/open-dynamic-robot-initiative/robot_interfaces/pull/34).

luator committed:
Add CheckpointTimer to simplify time measurement in loops

The `CheckpointTimer` class is intended for time measurement of loops
where the time spent in separate sections of the code is measured, as
well as the total duration of the loop.  See doxygen of the class for
more details.

It is basically a wrapper around a list of `Timer` instances with the
purpose to not clutter the measured code too much.  Also it can easily
be disabled via a template argument which should result in the compiler
optimizing away the function calls completely (i.e. no need to have
commented lines in the code).

MaximilienNaveau commented on include/real_time_tools/checkpoint_timer.hpp:
```diff
@@ -0,0 +1,86 @@
+ *
+ * Example:
+ *
+ * ~~~{.cpp}
```
This should be an actual code compiled. See @example in doxygen and in the ci_example/demos/ file.

MaximilienNaveau replied:
You can then copy the doc here (@codesnippet) for more details or reference to it (@see).

luator replied:
done

MaximilienNaveau commented on include/real_time_tools/checkpoint_timer.hpp:
```diff
@@ -0,0 +1,86 @@
+private:
+    std::array<real_time_tools::Timer, NUM_CHECKPOINTS + 1> timers_;
+    std::array<std::string, NUM_CHECKPOINTS + 1> checkpoint_names_;
+    size_t current_checkpoint = 1;
```
needs a trailing underscore.

luator replied:
done

MaximilienNaveau reviewed (commented):
Sorry I did not click on submit the review...

luator committed:
Fix name of variable and add missing doxygen

luator committed:
Move example to demo file

Move the example code to a complete demo file that is actually
compilable and runnable.  Then link to this file embed the example
snippet in the class documentation.
